merge v3: datajoint-style implementation (3/3)

ixcat opened the pull request:
see #2 , #3 
which were cancelled

Commit:
.gitignore: copy from djpy+emacs tweaks

Commit:
.gitignore: ~$ file tweak

Commit:
initial todo/file classification; copy in of items up to Subject

Commit:
reference.py: djify. currently included classes are loading.

Commit:
reference.py: merge orig-module/class comments under respective 'new' classes

Commit:
reference.py: misc format tweak.

Commit:
conversion as outlined in todo.txt up to subject

Commit:
subject.py: update id description fields

Commit:
reference.py: update description fields

Commit:
reference.py: fix up typo hmm...

Commit:
equipment.py: initial conversion of equipment related classes

Commit:
actions.py: initial conversion w/r/t autoconvert.

Commit:
actions.py; zap bookmark comment

Commit:
reference.py: add misc.Note and misc.CoordinateTransformation

Commit:
acquisition.py: initial conversion - django-vs-dj not quite lining up

Commit:
behavior.py: initial conversion implemented w/r/t acquisition DataDescription

Commit:
todo.txt: updated bookmarks.

Commit:
ephys.py: initial implementation using alyx DataSet model

Commit:
todo.txt: track progress

Commit:
imaging.py: initial conversion w/r/t Alyx' Dataset model

Commit:
todo.txt: document current state & outline issues discovered from 1st conversion pass.

Commit:
imaging.py: merge source-class comments into dj classes

Commit:
todo.txt: fmt misc

Commit:
reference.py: fmt nits

Commit:
alyx.manual1.py: import of 'combined' schema using Alyx' object model

Commit:
zap manual conversion from separate files

Commit:
alyx.manual1.py: misc tweaks

Commit:
alyx.manual1.py: misc tweaks

Commit:
alyx.manual1.py: flip schema object to use names mechanism

Commit:
alyx.manual1.py: flip schema object to use names mechanism

Commit:
Merge branch 'master' of github.com:ixcat/djcat-alyx

Commit:
git mv alyx.manual1.py alyx_manual1.py

Commit:
djified reimplementation thus far (->acquisition)

Commit:
alyx_manual1.py: fixup dj.Manual baseclass in COordinateTransformation

Commit:
alyx_manual1: add initial ipy notebook & erd png

Commit:
todo.txt: update w/r/t current state of affairs

Commit:
subject.py: initial datajoint-style implementation

Commit:
acquisition.py: initial datajoint-style implementation

Commit:
equipment.py: initial datajoint-style implementation

Commit:
actions.py: initial datajoint-style implementation

Commit:
alyx_dj.ipynb: initial add w/ current state erd plot

Commit:
actions.py: misc cleanup

Commit:
behavior.py: initial datajoint-style implementation

Commit:
behavior.py: note missing TimeScale in dj conversion

Commit:
equipment.py: add ProbeModel to equipment

Commit:
ephys.py: initial datajoint-style implementation

Commit:
alyx_dj.ipynb: add ephys to plots

Commit:
imaging.py: initial datajoint-style implementation

Commit:
update jupyter notebook and todo for complete dj-style implementation

Commit:
alyx_dj.ipynb: flip to full size prints

Commit:
png's for all\!

eywalker commented on actions.py:
```diff
@@ -0,0 +1,122 @@
+    # XXX: session_type table?
+    definition = """
+    -> subject.Subject
+    number:			integer		# number
```
better to prepend more specific keyword like `session_` - i.e. make it `session_number` rather than just `number`.

ixcat replied:
makes sense.

eywalker commented on actions.py:
```diff
@@ -0,0 +1,122 @@
+    -> subject.Subject
+    number:			integer		# number
+    ---
+    start_time:			datetime	# start time
```
similarly given that we'd likely want to join across tables, better to keep attributes as distinct as possible - make it `session_start_time` and `session_end_time`.

ixcat replied:
will do - only recently becoming aware of this issue

eywalker reviewed (changes requested):
Common attribute name is best avoided - instead of using `start_time` and `end_time` repeatedly, try to make them more specific to avoid collisions.

eywalker commented on behavior.py:
```diff
@@ -0,0 +1,103 @@
+
+
+@schema
+class Movie(dj.Manual):
```
relationship between `PupilTracking` and `Movie` is not super clear. Is Pupil tracking also a movie? If so I'm not sure that it makes sense to have a table `Movie` just to serve as a collection of all movies.

ixcat replied:
need to double check if these are stimulus videos or observation videos..
from memory it seems they were stimulus, so I broke out to prevent storing the same large blob in each record, esp. since the video could be reused among different experimental sessions

eywalker replied:
In case these are stimulus videos, I agree that it'd make sense to break it out into its own table. However, given the structure, the movie field of the pupil tracking is most likely referring to the vide of the pupil and hence it would make the most sense for the movie to be directly a blob (or more likely an external) inside the PupilTracking table.

ixcat replied:
Indeed this seems to be the the case -
I think was bogged down w/r/t 'TimeSeries', which in turn is a 'DataSet', and so saw the link
as a potentially 1:M relationship with respect to movies. updating accordingly.
```
    movie = models.ForeignKey(TimeSeries, blank=True, null=True,
                              related_name="pupil_tracking_movie",
                              help_text="Link to raw data")
```

eywalker commented on actions.py:
```diff
@@ -0,0 +1,122 @@
+
+
+@schema
+class Session(dj.Manual):
```
this makes more sense to be grouped with other experiment related tables rather than being "actions"

ixcat replied:
makes sense; to some extent mirrors upstream schema, will readjust

eywalker commented on behavior.py:
```diff
@@ -0,0 +1,103 @@
+    description:		varchar(255)	# description
+    """
+
+    class Time(dj.Part):
```
To me, it doesn't really make sense that `Time`, `Position` and `Power` are separate part tables. If they are just describing single optogenetic stimulus, they ought to be just fields in the `OptogeneticStimulus` table itself.

ixcat replied:
iirc these had a 1:M relationship - e.g. series of stimulation events.. but then why are they blobs. 
Hmm. potentially could be combined into 1 part table either way.. will audit w/r/t source

eywalker replied:
Yes, if they are 1:M relationship then this could make more sense. But even then, it's weird to me that they are separate 1:M - if anything it sounds like there should be single `Part` table (say `Pulse`) that then has an additional primary key attribute (to make for valid 1:M) and three dependent fields: time, position and power. Let's discuss this point.

ixcat replied:
xcheck w/r/t alyx has:
```
    stimulus_times = models.ForeignKey(Dataset, blank=True, null=True,
                                       related_name="optogenetic_stimulus_times",
                                       help_text="link to an n*2 array of start and stop "
                                       "of each pulse (sec)")
    stimulus_positions = models.ForeignKey(Dataset, blank=True, null=True,
                                           related_name="optogenetic_stimulus_positions",
                                           help_text="link to an n*3 array of stimulus positions")
    power = models.ForeignKey(Dataset, blank=True, null=True,
                              related_name="optogenetic_stimulus_power",
                              help_text="link to an n*1 array giving each pulse power")
```
so, will combine as ```Pulse``` per suggestion since this is in line with your thinking

eywalker commented on ephys.py:
```diff
@@ -0,0 +1,113 @@
+    ccf_lr:			float		# ccf lr
+    ---
+    allen_ontology:		varchar(255)	# allen ontology
+    json:			varchar(255)	# json
```
any idea what this field is supposed to hold? just generic information?

ixcat replied:
if ccf ap/dv/lr doesn't make sense to you, it doesn't make sense to me either :)

eywalker replied:
Ah, I actually was referring to the `json` field. The `ccf` stands for common coordinate framework - a specification to describe mouse brain coordinates established by Allen Institute (http://help.brain-map.org/display/mousebrain/API). `ap` = anterior-vs-posterior, `dv`=dorsal-vs-ventral (same as superior-vs-inferior in the picture, respectively) and `lr`=left-vs-right.

![image](https://user-images.githubusercontent.com/3084661/32133128-44db41bc-bb97-11e7-808a-d6850c085113.png)

ixcat replied:
ah yes, the original extended the base django object so that everything had a UUID pkey and json field, which were included in the machine translation. we are zapping in the 'dj-native' copy. Good catch - will grep for any other missed ones before updating.

eywalker commented on ephys.py:
```diff
@@ -0,0 +1,113 @@
+    # <class 'electrophysiology.models.ProbeModel'>
+    definition = """
+    -> actions.Session
+    start_time:                 datetime        # start time
```
Why is an `ExtraceCelluarRecording` identified by both `start` and `end` time? Also again, you have to rename the `start_time` and `end_time` for it to not collide with the same named PK attributes in `Session`.

ixcat replied:
renaming, ok.

as for concepts: what would be better w/r/t start/end?

to me, this allows easy indexing into other things than say, start time and then needing to add some computed thing based on length of other things, but not sure what offhand (had something in mind earlier)

eywalker replied:
To me it looks like the (renamed) `start_time` will be enough - needing `end_time` as part of PK attr suggests that two things can have identical `Session` and `start_time` but different `end_time`.

ixcat replied:
Could there not be 2 recordings in one session?
if so, perhaps using start+length might prevent the lack of clarity you are referring to?

I've updated to rename all start/end time attributes to be specific to the type; current list of all start/end attributes are as follows:

```
$ egrep '(start|end)_time' *.py |grep -v alyx
actions.py:    surgery_start_time:              datetime        # surgery start time
actions.py:    surgery_end_time:                datetime        # surgery end time
actions.py:    session_start_time:         datetime     # start time
actions.py:    session_end_time:           datetime     # end time
actions.py:    restriction_start_time:     datetime     # start time
actions.py:    restriction_end_time:       datetime     # end time
actions.py:    other_action_start_time:    datetime     # start time
actions.py:    other_action_end_time:      datetime     # end time
behavior.py:    pupil_tracking_start_time:  datetime        # start time
behavior.py:    head_tracking_start_time:   datetime        # start time
behavior.py:    administration_start_time:  float               # start time
behavior.py:    administration_end_time:    float               # end time
ephys.py:    recording_start_time:       datetime        # start time
ephys.py:    recording_end_time:         datetime        # end time
imaging.py:    widefield_start_time:       datetime     # nominal start time
imaging.py:    widefield_end_time:         datetime     # nominal end time
imaging.py:    2p_start_time:              datetime     # nominal start time
imaging.py:    2p_end_time:                datetime     # nominal end time
imaging.py:    roi_start_time:             datetime     # nominal start time
imaging.py:    roi_end_time:               datetime     # nominal end time
``` 
would you advise as to which of these (other than session) might be extraneous?

or should I assume previous comment is comprehensive w/r/t this sort of attribute  throughout existing schema, and only this table needs updating?

ixcat replied:
clarified was only for pkey attributes, of which this was only occurrence.

eywalker commented on equipment.py:
```diff
@@ -0,0 +1,101 @@
+    # <class 'equipment.models.VirusBatch'>
+    definition = """
+    -> Supplier
+    batch_id:			int             # batch id
```
I'd think that `VirusBatch` should be characterized by the `virus_type` as well?

ixcat replied:
makes sense. will shift this up.

eywalker commented on equipment.py:
```diff
@@ -0,0 +1,101 @@
+
+
+@schema
+class Appliance(dj.Manual):
```
Not too clear what's gained by grouping all items under a single `Appliance` table. How natural is it that all "appliances" are enumerate by `appliance_id`?

ixcat replied:
No idea. not sure why lab location is relevant either. This is one copied from Alyx, which used inherited classes here. Could split it, Dimitri didn't seem to mind it or actually liked when we spoke previously.

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+class Species(dj.Manual):
+    # <class 'subjects.models.Species'>
+    definition = """
+    species_id:			int             # species id
```
given binomial info will be inserted, using arbitrary `species_id` will only make downstream tables harder to use. I suggest replacing `species_id` with `binomial`

ixcat replied:
makes sense

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+class Strain(dj.Manual):
+    # <class 'subjects.models.Strain'>
+    definition = """
+    strain_id:			int             # strain id
```
Same thing here - unless they mean to use some unique but descriptive shorthand as the `strain_id` (probably not the case given it's an `int`), it would be better to have `descriptive_name` be the primary key. I'd rename it to `strain_name` or something along that line.

ixcat replied:
yep. i think for some of these id types I just copied in on the side of caution since it doesn't pollute so much w/r/t classes.. but in any event, will do

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+class Sequence(dj.Manual):
+    # <class 'subjects.models.Sequence'>
+    definition = """
+    sequence_id:		int             # sequence id
```
technically this should be identified by the `base_pairs` but that's understandably inconvenient - if they have a better short name alternative (e.g. if their `informal_name` can be unique) then I'd suggest that be the primary key. Otherwise, it'll be fine to leave this identified by `sequence_id`.

ixcat replied:
will xref to see if the informal name was infact unique upstream and adjust accordingly.

ixcat replied:
ah in fact 'informal_name' was being used but via a django mechanism I didn't initially recognize as being a pkey. adjusted.

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+
+
+@schema
+class Subject(dj.Manual):
```
This appears to be too heavily populated with attributes. It'd make sense to break off some attributes into its own table, especially given it is unlikely that every attribute will be filled for every mouse (even if that's the case eventually, I'd imagine many mouse will not have i.e. implant for some period of time).

Also markers like `to_be_genotype` is not a good attribute as that'd expected to change. Rather, there should be a table like `Genotype` that points to `Subject` and one should determine if the mouse needs to be genotyped or not by checking `Genotype & Subject` presence.

ixcat replied:
will zap the state tracking fields - as for attributes, any pointers on what to break down where would be welcome. seemed alot to me as well, but wasn't sure what/where to reorganize.

eywalker replied:
Yep - let me look through a bit more in detail and come back with suggestion.

eywalker commented on ephys.py:
```diff
@@ -0,0 +1,113 @@
+
+
+@schema
+class SpikeSorting(dj.Manual):
```
Instead of naming a table after the algorithm/action, let's name it to describe the entity within the table. Here, I'd suggest renaming to something like `SortedUnitGroup`.

ixcat replied:
done.

eywalker commented on equipment.py:
```diff
@@ -0,0 +1,101 @@
+    # ... assuming Alyx' separate classes vs field is reasoned.
+    definition = """
+    -> LabLocation
+    -> EquipmentModel
```
To me, it appears that `EquipmentModel` specifies exactly what kind of appliance this is already and thus use of specific part tables feel a bit redundant. If anything, breakdown into specific object category ought to be done at the level of `EquipmentModel` (i.e. `EquipmentModel` can have part tables like `WeighingScale` etc) instead.

ixcat replied:
perhaps merging EquipmentModel fields into Appliance and zapping EquipmentModel makes more sense?

eywalker commented on reference.py:
```diff
@@ -0,0 +1,114 @@
+
+
+@schema
+class User(dj.Manual):
```
Quite clearly, this table is used as backend table for Alyx system and thus many of the fields do not necessary make sense as scientific schema. We should discuss how should "users" in the sense of the system user should (or should not) be separated from "users" in the sense of experimenters.

ixcat replied:
true, this could be django baggage (users are application level), leaving as is for now pending further discussion

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    target_phenotype:			varchar(255)	# target phenotype
+    auto_name:				varchar(255)	# auto name
+    is_active:				boolean		# is active
+    subject_autoname_index:		integer		# subject autoname index
```
I have no idea what these `auto_name` fields refer to...

ixcat replied:
this had something to do with tracking indexes into key generation code logic; possibly can just be zapped

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    -> Line
+    sex:			enum("M", "F", "U")	# sex
+    birth_date:			date			# birth date
+    death_date:			date			# death date
```
Inclusion of field that has to be added later is typically a bad idea as this implicitly requires use of `UPDATE` on the table. Let's separate `death_date` into separate `Manual` table.

ixcat replied:
should birth/death be separate parts of Subject?

eywalker replied:
`birth_date` can stay as part of the subject but death date will be better as a separate table (and not a `Part` table). This is because you would usually not know the death date of the subject at the time of the subject entry whereas birthday is usually known (or if not known can remain empty - depending on how strict the designer wants this field to be filled).

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    sex:			enum("M", "F", "U")	# sex
+    birth_date:			date			# birth date
+    death_date:			date			# death date
+    wean_date:			date			# wean date
```
Same thing with `wean_date` - these are information that's not going to be known at the time of the mouse data entry, and should be separated into its own `Manual` table.

ixcat replied:
as above;

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    cull_method:		varchar(255)		# cull method
+    adverse_effects:		varchar(255)		# adverse effects
+    (actual_severity)		-> reference.Severity   # actual severity
+    to_be_genotyped:		boolean			# to be genotyped
```
Let's remove `to_be...` fields.

ixcat replied:
done; agreed this can be computed if it is needed and plus it implies state changes which is not in keeping with preference for immutable records

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    reduced_date:		date			# reduced date
+    """
+
+    class Genotype(dj.Part):
```
Given that genotyping of mouse is something that's typically done after mouse info is already entered into the database, it really shouldn't be `Part` table. Let's separate `Genotype` into its own `Manual` table, but keeping the same definition.

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    (request)                   -> SubjectRequest(subject_request_id)
+    -> Source
+    lamis_cage:			integer			# lamis cage
+    implant_weight:		float			# implant weight
```
Similarly, not every single mouse goes through implantation and thus it makes no sense for this field to be present in every mouse. Let's create another table called `Implant` that references `Subject` as parent with non PK attr of `implant_weight` and perhaps `implant_description: varchar`

ixcat replied:
done

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    (responsible_user)          -> reference.User
+    (request)                   -> SubjectRequest(subject_request_id)
+    -> Source
+    lamis_cage:			integer			# lamis cage
```
I have no idea what `lamis` stands for but I'm guessing that this is referring to a specific cage that the mouse is placed in. Assuming that this guess is correct, then again this is something that can vary over the life course of mouse, and it's a bad idea to include this in the `Subject` table. If anything, we should make `Caging` table that has `cage_id` and `-> Subject` and most likely `timestamp`  as primary key. As long as they would make a new entry into this table every time a mouse is placed into a new cage, the table will allow them to keep track of which mouse were in which cage.

ixcat replied:
done

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+    ear_mark:			varchar(255)		# ear mark
+    protocol_number:		varchar(255)		# protocol number
+    description:		varchar(255)		# description
+    cull_method:		varchar(255)		# cull method
```
Anything related to culling should be moved into it's own (set of) tables.

ixcat replied:
done.

eywalker commented on subject.py:
```diff
@@ -0,0 +1,178 @@
+
+
+@schema
+class Birth(dj.Manual):
```
I feel that table name like `BreedingPair` will in fact be more informative.

ixcat replied:
Dimitri didn't like 'BreedingPair' - If I recall correctly it is because the parents are a property of the subject itself additionally there is the 3rd 'parent' attribute (currently disabled but likely needed) I am open to suggestions but not changing initially since your opinions conflict here

Commit:
action.py: update session attirbutes per code review

Commit:
ephys.py: zap leftover json field

Commit:
reference.py: zap leftover 'json' field

eywalker commented on equipment.py:
```diff
@@ -0,0 +1,101 @@
+    definition = """
+    -> LabLocation
+    -> EquipmentModel
+    appliance_id:		int             # appliance id
```
Now I believe that `appliance_id` is used to demarcate multiple instances of the same appliance in a particular location. Not exactly sure what value would go in there, but probably it's just an enumeration from `0, 1, 2,...`

ixcat replied:
zapping. serial number should be unique.

Commit:
behavior.py: combine OptogeneticStimulus part tables into single 'Pulse' table

Commit:
behavior.py: clarify OptogeneticStimulus.Pulse blob field usage

Commit:
behavior.py: zap 'Movie' table since these are actually table attributes

Commit:
{actions,behavior,ephys,imaging}.py: fix *_{date,start,stop} collisions

Commit:
actions.py: aend_time -> end_time

Commit:
ephys.py: move recording_end_time out of pkey

Commit:
equipment.py: move VirusBathc.virus_type into pkey

Commit:
subject.py: zap Species.species_id and use Species.binomial as pkey

Commit:
subject.py: zap Strain.strain_id and use Strain.strain_name as pkey

Commit:
subject.py: zap Sequence.sequence_id and use Sequence.sequence_name as pkey

Commit:
ephys.py: rename SpikeSorting to SortedUnitGroup

Commit:
subject.py: move changing Subject attributes to tables

Commit:
subject.py: caging_date: time -> datetime

Commit:
equipment.py: rearrange Appliance classes per feedback

Commit:
subject.py: refactored Genorype table should be dj.Manual

Commit:
actions.py: fix reference to refactored WeighingScale

Commit:
behavior.py: fix/clarify Appliance reference to refactored LightSource

Commit:
ephys.py: fix references to refactored equipment classes

Commit:
imaging.py: fix equipment refs, 2p_<attr> -> twop_<attr>

Commit:
re-run jupyter notebook with new changes

Commit:
subject.py: zap 'autoname index' fields

Commit:
{actions,behavior}.py: move Session into behavior schema:

Commit:
update jupyter & images to reflect latest

ixcat commented:
I believe all items are accounted for in latest.

Commit:
subject.py: rename implanting to implant

Commit:
behavior.py: make session_start_time a pkey attr

Commit:
behavior.py: rename head/pupil tracking from x_y_<> to descriptive name.

Commit:
behavior.py: OptogeneticStimulus: flatten to 1 record per w/timestamp.

Commit:
behavior.py: Pharmacology: make volume/concentration into float

Commit:
ephys.py: clarify/add {intra,extra}cellular_rec_{start,end} attributes

Commit:
imaging.py: drop WidefieldImaging.raw_frames NULL default.

Commit:
imaging.py: make ROIDetection a dj.Computed table.

Commit:
ephys.py: make BrainLocation a SortedUnitGroup.Unit attribute instead of pkey

Commit:
zap actions.py -> move classes to subject.py

Commit:
{ephys,reference}.py: move BrainLocation into reference module

Commit:
alyx: zap many 'User' references

Commit:
reference.py: zap usergroups/permissions

Commit:
{acquisition,behavior}.py: move behavior into acquisition & zap behavior

... these classes were more acquisition related...

also, previous 'DataDescription' file ingest provenance class has been removed.

Commit:
subject.py: cleanup ingest

Commit:
subject.py: cleanup ingest

Commit:
{ephys,imaging}.py: flip behavior to acquisition

Commit:
update of notebook&images w/r/t recent dev
